### ISSUE TITLE
Ensure compatibility between Pinterest Hover button and lazy loading image plugins

### ISSUE THREAD

**Issue by Nicholas-Cardot** [warfare-plugins/social-warfare-pro#13]
Lazy Loading Problem
-Lazy Loading doesn't work the Pinterest image hover button activate 

Lazy Loading Solution
-Disable Pinterest image hover button

_________________________________

Part of this is because the img URL is stored in a different parameter prior to being revealed on the screen. The solution is most likely to identify that other attribute and target it instead of only the src attribute. I have already implemented this in a recent update. Let's review the modifications I've already made and test it prior to taking further action. You can see them here:

https://github.com/warfare-plugins/social-warfare/blob/master/js/script.js#L364

Line 364 - 370

**Comment by Nicholas-Cardot** [comment]
**Notes for Our Developer**

1. Install Jetpack and activate their lazyload feature.

2. Create a blog post with numerous images especially down below the fold so that the lazy load animation can be observed.

**Important Note:** Remember that you'll be working on the unminified version of the /js/script.js file. In order to get that version to load instead of the /js/script.min.js file, you'll need to add ?swp_debug=script to the blog post URL.

3. Check to ensure that our pin hover image is being properly applied to the images that are lazyloaded. Right now our script targets the src attribute of the image to create the Pinterest link. On lazy loaded images, the src attribute is often stripped out and stored in a "media" or some other attribute. Find what attribute the plugin uses and make sure that our script takes it into account when creating the pin button.

4. Install WP Rocket and activate their lazy load feature.

5. Repeat steps 2 and 3 with the new plugin.

6. Write any notes, thoughts or observations in the comment section here on this thread.

7. When finished, be sure to create a pull request so that I can review it and pull it back in.

**Comment by scott-wiser** [comment]
1)Jetpack doesn't have some features on private servers, even in dev mode.

Do you have a public url with Jetpack installed on I can access and test changes on? 

If not, I have a public hosting I can use, it's just not setup for that yet.  Jetpack doesn't have some features on private servers, even in dev mode.

4) Do you happen to have WP Rocket Pro with unlimited licenses, or can I get a license for the Pro if that needs to be tested?

**Comment by scott-wiser** [comment]
Although I don't have a WP Rocket Pro license, the following points, along with the documentation link answers some questions.

1) Before the images are fully loaded, WP Rocket Pro uses a data-lazy-src attribute for the image url, with the src containing a default statement.  Assumedly, the lazy load script changes the src attribute from the data-lazy-src attribute.

So, with Javascript, we should be able to modify, or use, the img element through identifying the data-lazy-src attribute, which should contain the image url.

2) With PHP, we can detect if the WP Rocket Pro is going to use lazy loading through the PHP functions  function_exists( 'get_rocket_option' ) and get_rocket_option( 'lazyload' ).  If these are both true, then it's likely a WP Rocket lazy load will happen, and so the could can adapt for compatibility there.

http://docs.wp-rocket.me/article/130-manually-apply-lazyload-to-an-image

**Comment by Nicholas-Cardot** [comment]
Have you looked at the function in script.js? What do you think is the best course of action?

**Comment by scott-wiser** [comment]
It looks like the data-lazy-src should work ok for WP Rocket Pro.  I can test if you have a Pro version and want another tester of the WP Rocket Pro.

For the Jetpack I'll have to use a public url and setup a test environment (unless you have one available for testing)

Also, how about backwards compatibility?  How far back should we verify for plugin versions of Jetpack and WP Rocket Pro?

**Comment by Nicholas-Cardot** [comment]
Only the current version. We don't support outdated plugins or themes.

**Comment by scott-wiser** [comment]
From a code review, it looked as if the WP Rocket Pro was ready to go.  Let me know if you have  a way to test Pro or if I should get a license for it, or if you have one for me.

I'll see about getting public hosting setup to test Jetpack if you don't have a test environment ready.  The lazy load won't work on private network test environments.

I'll have more tomorrow on the results

**Comment by nutsandbolts** [comment]
I'll send you a link on Slack to download WP Rocket.

**Comment by GeekPress** [comment]
Hi Scott,

Jonathan from WP Rocket.

If you need to have a free license to be able to test WP Rocket, please contact us at contact[at]wp-rocket.me

We will be happy to provide you a free license to be able to fix the issue.

Kind Regards

**Comment by scott-wiser** [comment]
**WP Rocket Lazy Load**
Hi Jonathan, thanks.

I was able to test the lazy load feature.  The Social Warfare Pinterest save hover feature works fine, as I suspected when comparing it with the documentation on the WP Rocket site.  The documentation was helpful.

**JetPack Photon Feature**
I also tested the Jetpack.  On it's own, with the Photon feature enabled, the images load with the Pinterest save hover feature.

**Both JetPack Photon and WP Rocket Lazy Load**
With both JetPack Photon and WP Rocket Lazy Load enabled at the same time, the images do not load with the Pinterest save hover feature.

For some reason, using both as the same time causes a conflict which does enable the images to load, just not the Pinterest save hover feature.

Temporary Solution:
If both the JetPack Photon and WP Rocket Lazy Load are enabled at the same time, disabling one feature or another might solve the problem.  In my test case, admin do not appear to have to disable the entire JetPack or WP Rocket plugins, simply the Lazy Load or Photon feature.

My hopes are most systems will behave similarly to this test case.  I'm using:
JetPack Version 4.7.1
WP Rocket Version 2.9.8.1
Wordpress Version 4.7.3
Social Warfare Pro 2.2.3

Long Term Solution:
A long term solution would involve measuring which attributes appear with both features enabled, and attempting to modify the code to still get it right.  This might vary between version combinations, although there might be an innovative solution that will prevent the problem.

We could also detect which features are enabled with PHP, and modify Javascript variables to reflect that and have the Javascript account for the differences.

I'm not sure on the time costs for a long term solution, or whether it would work with all versions, so I've left the temporary solution and will wait for feedback
